### DISCUSSION TITLE
RandFlip Spatial Axis and Probabilities both as Sequences

### DISCUSSION THREAD

**Discussion by AustinTapp** [Project-MONAI/MONAI#5356] [Ideas]
For RandFlip, spatial_axis can be set individually (i.e. [0], [1], [2], for x, y and z respectively) or spatial_axis can be set as a sequence[int] (i.e. [0, 1, 2], for x, y and z respectively). However, probabilities can only be set as a single float. 

It would be helpful to allow probability arguments to also be passed in the same manner as spatial_axis: _Optional[Union[Sequence[int], int]] = None_. This would permit users to define a sequence, where the probability float matches the corresponding spatial_axis in a sequence. The typical behavior, which uses a single int for all random flip probabilities, should be maintained.

For example:
RandFlipd(keys=["keys"], spatial_axis=[0, 1, 2], prob=[0.1, 1, 0.5])
Would flip an image along the x-axis 10%, along the y-axis 100%, and along the z-axis 50% of the time. Note that this definition would be for independent flipping probability. 

Currently, users must specify multiple RandFlip transforms for differing axes if they would like to initialize each flip independently. An example of this occurs in https://github.com/Project-MONAI/tutorials/blob/main/self_supervised_pretraining/ssl_finetuning_train.py#L87

Thanks for your consideration and support.

**Comment by KumoLiu** [comment]
Hi @AustinTapp, the `spatial_axis ` set in this sequence form, means flip on multiple dimensions at the same time, not flip on them separately. Like the example from [PyTorch's doc](https://pytorch.org/docs/stable/generated/torch.flip.html).
```
x = torch.arange(8).view(2, 2, 2)
x --> tensor([[[ 0,  1],
              [ 2,  3]],
      
              [[ 4,  5],
               [ 6,  7]]])
torch.flip(x, [0, 1])
x --> tensor([[[ 6,  7],
               [ 4,  5]],
  
              [[ 2,  3],
               [ 0,  1]]])
```
Hope it can help you, thanks!

**Comment by AustinTapp** [comment]
Hi @KumoLiu,

Thanks for your quick reply.

I understand `spatial_axis` can be set as a sequence. However, when providing a probability, only a single float is specified. Therefore, if the flip does occur, it will be performed along ALL the axes specified in the sequence. 

Instead, I suggest users could **additionally** provide their probability as a sequence. The `prob` sequence indices would match the sequence indices of `spatial_axis`. With this notation, each flip would be performed independently with an independent probability given by the matching index.

Without this, users must specify multiple `RandFlip` commands if they would like independent axis flipping with independent probabilities for each, as in: https://github.com/Project-MONAI/tutorials/blob/main/self_supervised_pretraining/ssl_finetuning_train.py#L87

Best regards,
Austin

**Reply by KumoLiu** [reply]
Hi @AustinTapp, thanks for your advice.
If we make probability to a sequence, for your example `RandFlipd(keys=["keys"], spatial_axis=[0, 1, 2], prob=[0.1, 1, 0.5])`, it may be confusing whether we want to flip in all three directions at the same time or flip in a certain direction, and if we want to flip in [x, y] plane, we also need to specify another `RandFlipd`, right?
Thanks!

**Reply by AustinTapp** [reply]
Hi @KumoLiu,
I believe the existing function is set to flip in all specified directions at the same time by only passing `prob` as as a single float. I think this is useful and should be maintained so that `RandFlipd(keys=["keys"], spatial_axis=[0, 1, 2], prob=1.0)` has a 100% chance to randomly flip the image in all three directions [x, y, z] at once.

If you want to flip in the [x, y] plane at the same time, then you can already specify those two axes as `RandFlipd(keys=["keys"], spatial_axis=[0, 1], prob=1)` - you would not need another `RandFlipd`. 

But if you want the image to _possibly_ flip in both the [x, y] plane, where there is a chance that _both_ flips do not happen then you can say `RandFlipd(keys=["keys"], spatial_axis=[0, 1], prob=[1, 0.5])`. In this case, the image flipping probabilities for each axis occur independently. There is a 100% chance the image will be flipped in the x direction, then that flipped image also has a 50% chance of being flipped in the y direction. Thus, 50% of the time you will get an image flipped in the [x, y] direction and 50% of the time you will get an image flipped **only** in the [x] direction. 

Currently, for this to be achieved, a user must run: `RandFlipd(keys=["keys"], spatial_axis=[0], prob=1)` and then `RandFlipd(keys=["keys"], spatial_axis=[1], prob=[0.5])`.

Best regards,
Austin
